flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/logical/CalcUpsertToRetractionTransposeRule.scala

```diff
@@ -0,0 +1,109 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.plan.rules.logical
+
+import org.apache.calcite.plan.RelOptRule.{none, operand}
+import org.apache.calcite.plan.{RelOptRule, RelOptRuleCall}
+import org.apache.calcite.rex.{RexCall, RexInputRef}
+import org.apache.calcite.sql.SqlKind
+import org.apache.flink.table.plan.nodes.logical.{FlinkLogicalCalc, FlinkLogicalUpsertToRetraction}
+
+import scala.collection.JavaConversions._
+
+/**
+  * Use this rule to transpose Calc through UpsertToRetraction relnode. It is beneficial if we get
+  * smaller state size in upsertToRetraction.
+  */
+class CalcUpsertToRetractionTransposeRule extends RelOptRule(
+  operand(classOf[FlinkLogicalCalc],
+    operand(classOf[FlinkLogicalUpsertToRetraction], none)),
+  "CalcUpsertToRetractionTransposeRule") {
+
+  override def matches(call: RelOptRuleCall): Boolean = {
+    val calc = call.rel(0).asInstanceOf[FlinkLogicalCalc]
+    val upsertToRetraction = call.rel(1).asInstanceOf[FlinkLogicalUpsertToRetraction]
+
+    // column pruning or push Filter down
+    calc.getRowType.getFieldCount <= upsertToRetraction.getRowType.getFieldCount &&
+    // key fields should not be changed
+      fieldsRemainAfterCalc(upsertToRetraction.keyNames, calc)
+  }
+
+  override def onMatch(call: RelOptRuleCall) {
+    val calc = call.rel(0).asInstanceOf[FlinkLogicalCalc]
+    val upsertToRetraction = call.rel(1).asInstanceOf[FlinkLogicalUpsertToRetraction]
+
+    // get new calc
+    val newCalc = calc.copy(calc.getTraitSet, upsertToRetraction.getInput, calc.getProgram)
+    // get new upsertToRetraction
+    val oldKeyNames = upsertToRetraction.keyNames
+    val newKeyNames = getNamesAfterCalc(oldKeyNames, calc)
+
+    val newUpsertToRetraction = new FlinkLogicalUpsertToRetraction(
+      upsertToRetraction.getCluster,
+      upsertToRetraction.getTraitSet,
+      newCalc,
+      newKeyNames)
+
+    call.transformTo(newUpsertToRetraction)
+  }
+
+  private def fieldsRemainAfterCalc(fields: Seq[String], calc: FlinkLogicalCalc): Boolean = {
+    // get input output names
+    val inOutNames = getInOutNames(calc)
+    // contains all fields
+    inOutNames.map(_._1).containsAll(fields)
+  }
+
+  /**
+    * Get related output field names for input field names for Calc.
+    */
+  private def getInOutNames(calc: FlinkLogicalCalc): Seq[(String, String)] = {
```

Comment:
Some kind of `map` usually better expresses mappings between two things compared to list of tuples. Is there a reason that I do not see why have you chosen list of tuples?

Reply:
ditto